### ISSUE TITLE
Feature request: Watchdog setting

### ISSUE THREAD

**Issue by Cairan** [KSP-KOS/KOS#2053]
Sometimes, a probe's code will crash due to unforeseen circumstances. A "watchdog" setting could be provided so that the kOS CPU core either:

a) Does nothing different than current behavior
b) On normal program termination, reboots the CPU and reloads the boot file.
c) On both normal and abnormal terminations, reboots the CPU and reloads the boot file

This could either be set via the UI or via a processor property / suffix related to the boot file options

This watchdog feature is a real life one, providing a safe mode for both hardware and software issues on current operational space probes and satellites.

**Comment by Dunbaratu** [comment]
The danger with "on crash, reboot" is what happens when the boot script *IS* the cause of the crash.  How do you break out of the cycle?

**Comment by Cairan** [comment]
I see your point. In that case the watchdog could be on duty for anything but the boot script. Or we could have a watchdog-triggered reboot counter. By default the max number of watchdog reboots could be set to some value (a few reboots) but overriden if desired... by keeping such a counter between reboots the user could also switch to an alternate program if so desired if the main one it keeps crashing and the boot script is only checking the counter and launching a daughter program based on that value... ie. switch to a safe mode program.

**Comment by zegkljan** [comment]
I love this idea! I would say, if your boot script crashes, it's your problem :).

However, I could see another way of handling crashes - have a crash handler/safe mode program, i.e. a program that is run if and only if a crash happens in the main program. It could be set in the same way as a boot file, i.e. when building a vehicle, you would choose your crash handler. If it is set, a crash starts up this handler, if it is not set, a reboot happens. Or, alternatively, crash handler is always run and it is up to you whether you just want to reboot or do something else. If there is no crash handler, the crash is not handled.

The "always run crash handler" approach could also handle the difference between normal exit and an actual crash - the handler could be expected to take one parameter - a flag indicating whether a crash happened or the program exited normally. This could be extended even more, e.g. by sending a lexicon (or a new special structure just for this purpose) that would describe the circumstances under which the crash handler was started.

Another way that I see is to add a new suffix to the processor that would indicate whether which state is it in - normal, after crash, after exit (again, potentially holding more information) - and reboot, keeping this state so that the boot script can decide.

But however would this be implemented, I'd love this.

**Comment by Cairan** [comment]
I love the idea of a "normal boot" script parameter and a "safe mode" script parameter, with the handler calling the safe mode if such a script is defined, just like the boot option is handled by kOS on power up / reboot. That whould require the least amount of fidling with the existing logics. Your idea of a lexicon being passed as a parameter for the safe mode is great too ...

**Comment by mgalyean** [comment]
A bit late here, but a boot->crash->reboot->crash is very realistic.  What does one do in the real world in this case?  You shutdown the power, and try to fix the boot sector by booting from another media.  I don't think a boot/crash cycle is reason enough to not consider allowing it, as long as the plug can be pulled and the boot file can fixed.  So maybe as mentioned above, a "safe" boot mode that will always give one a Proceed> prompt such that copypath could be used, etc

**Comment by Dunbaratu** [comment]
@mgalyean perhaps a way on the PAW during flight for the user to unset the bootfile (putting ``core:bootfilename`` back to ``""`` again) might do what you're saying?

**Comment by mgalyean** [comment]
Absolutely.  Clear it, or disable it, or somehow getting it to boot from backup media, like that would be great realism as it is the first thing crew would do with a malf'ing piece of equipment.  Maybe for career mode crew would have to be on board?  idk, but I like the clearing as that would probably be easiest to implement given the current PAW
